contracts/libraries/ControllerV2.sol

```diff
@@ -0,0 +1,233 @@
+import {IController} from "../external/clober-v2/IController.sol";
+import {IBookManager} from "../external/clober-v2/IBookManager.sol";
+import {BookId, BookIdLibrary} from "../external/clober-v2/BookId.sol";
+import {CurrencyLibrary, Currency} from "../external/clober-v2/Currency.sol";
+import {IControllerV2} from "../interfaces/IControllerV2.sol";
+import {SubstituteLibrary} from "./Substitute.sol";
+
+import {Epoch} from "./Epoch.sol";
+
+abstract contract ControllerV2 is IControllerV2, ERC1155Holder, Ownable2Step, ReentrancyGuard {
+    using SafeCast for uint256;
+    using BookIdLibrary for IBookManager.BookKey;
+    using SafeERC20 for IERC20;
+    using CouponKeyLibrary for CouponKey;
+    using CouponLibrary for Coupon;
+    using CurrencyLibrary for Currency;
+    using SubstituteLibrary for ISubstitute;
+
+    IWrapped1155Factory internal immutable _wrapped1155Factory;
+    IController internal immutable _cloberController;
+    ICouponManager internal immutable _couponManager;
+    IBookManager internal immutable _bookManager;
+    IWETH9 internal immutable _weth;
+
+    mapping(uint256 couponId => IBookManager.BookKey) internal _couponSellMarkets;
+    mapping(uint256 couponId => IBookManager.BookKey) internal _couponBuyMarkets;
+
+    constructor(
+        address wrapped1155Factory,
+        address cloberController,
+        address bookManager,
+        address couponManager,
+        address weth
+    ) Ownable(msg.sender) {
+        _wrapped1155Factory = IWrapped1155Factory(wrapped1155Factory);
+        _cloberController = IController(cloberController);
+        _couponManager = ICouponManager(couponManager);
+        _bookManager = IBookManager(bookManager);
+
+        _couponManager.setApprovalForAll(address(_cloberController), true);
+        _weth = IWETH9(weth);
+    }
+
+    modifier wrapAndRefundETH() {
+        bool hasMsgValue = address(this).balance > 0;
+        if (hasMsgValue) _weth.deposit{value: address(this).balance}();
+        _;
+        if (hasMsgValue) {
+            uint256 leftBalance = _weth.balanceOf(address(this));
+            if (leftBalance > 0) {
+                _weth.withdraw(leftBalance);
+                (bool success,) = msg.sender.call{value: leftBalance}("");
+                require(success);
+            }
+        }
+    }
+
+    modifier wrapETH() {
+        if (address(this).balance > 0) _weth.deposit{value: address(this).balance}();
+        _;
+    }
+
+    function _executeCouponTrade(
+        address user,
+        address token,
+        Coupon[] memory couponsToMint,
+        Coupon[] memory couponsToBurn,
+        int256 interestThreshold
+    ) internal {
+        uint256 length = couponsToBurn.length + couponsToMint.length;
+        IController.Action[] memory actionList = new IController.Action[](length);
+        bytes[] memory paramsDataList = new bytes[](length);
+        address[] memory tokensToSettle = new address[](length);
```

Comment:
`tokensToSettle` is an empty array. Does this logic appropriately run?

Reply:
Resolved.